client/jetpack-onboarding/main.jsx

```diff
@@ -69,12 +69,19 @@ class JetpackOnboardingMain extends React.PureComponent {
 	}
 
 	retrieveOnboardingCredentials() {
-		const { isConnected, siteId, siteSlug } = this.props;
+		const { isConnected, isRequestingWhetherConnected, jpoAuth, siteId, siteSlug } = this.props;
 		const { hasFinishedRequestingSite } = this.state;
+		const isMissingCredentials =
+			! isRequestingWhetherConnected && ! hasFinishedRequestingSite && ! jpoAuth;
 
 		// If we are not connected and missing the Jetpack onboarding credentials,
 		// redirect back to wp-admin so we can obtain them again.
-		if ( hasFinishedRequestingSite && ! isConnected && ! siteId && siteSlug ) {
+		if (
+			( hasFinishedRequestingSite || isMissingCredentials ) &&
```

Comment:
`isMissingCredentials` includes a `&& ! hasFinishedRequestingSite` criterion. Looks like we can simplify?

Reply:
Also, `hasFinishedRequestingSite` only depends on `isRequestingWhetherConnected` (flipping from `true` to `false`)...

Reply:
Thinking some more about this: Why is the issue actually happening? I.e. what's different from first landing at `/jetpack/start/<JPsite>`, which we know fetches JPO credentials correctly?

Reply:
Possibly related: https://github.com/Automattic/wp-calypso/issues/23435#issuecomment-374643089

Reply:
TBH I really didn't want to overcomplicate the logic here. #23472 looks promising, will have a look at it now.